### ISSUE TITLE
zip are unzipped

### ISSUE THREAD

**Issue by lecorguille** [galaxyproject/galaxy#9530]
I'm aware that zip isn't very welcome within Galaxy :)

But we still need them for example for NMR Data. The structure of one sample is rather complex.
<img width="963" alt="Capture d’écran 2020-03-17 à 10 33 08" src="https://user-images.githubusercontent.com/11072022/76842557-bdcde280-683a-11ea-96ec-a0abca616ab6.png">

So that's why I create this nasty but practical 
 no_unzip datatype: https://toolshed.g2.bx.psu.edu/view/lecorguille/no_unzip_datatype/7800ba9a4c1e

Since the 19.01 : 
 - the auto-detection still trigger the unzip and finally keep only one file
 - if we force no_unzip: 
```
Traceback (most recent call last):
  File "/shared/mfs/data/galaxy/server/tools/data_source/upload.py", line 326, in <module>
    __main__()
  File "/shared/mfs/data/galaxy/server/tools/data_source/upload.py", line 319, in __main__
    metadata.append(add_file(dataset, registry, output_path))
  File "/shared/mfs/data/galaxy/server/tools/data_source/upload.py", line 128, in add_file
    convert_spaces_to_tabs=dataset.space_to_tab,
  File "/shared/mfs/data/galaxy/server/lib/galaxy/datatypes/upload_util.py", line 70, in handle_upload
    if check_content and hasattr(datatype, 'sniff') and not datatype.sniff(path):
  File "/shared/mfs/data/galaxy/mutable-data/shed_tools/toolshed.g2.bx.psu.edu/repos/lecorguille/no_unzip_datatype/7800ba9a4c1e/no_unzip_datatype/no_unzip_datatypes.py", line 27, in sniff
    zf = zipfile.ZipFile(filename)
  File "/shared/mfs/data/software/miniconda/envs/_galaxy_/lib/python2.7/zipfile.py", line 770, in __init__
    self._RealGetContents()
  File "/shared/mfs/data/software/miniconda/envs/_galaxy_/lib/python2.7/zipfile.py", line 811, in _RealGetContents
    raise BadZipfile, "File is not a zip file"
zipfile.BadZipfile: File is not a zip file
```

I solved that with a quick and dirty patch: 
https://github.com/galaxyproject/galaxy/blob/dev/tools/data_source/upload.py#L97
```
auto_decompress = dataset.get('auto_decompress', True)
```
In:
```
auto_decompress = False
```

But I don't want to do that in our new usegalaxy.fr
Do you have a better workaround ?

**Comment by lecorguille** [comment]
A little up on this issue :/

**Comment by joachimwolff** [comment]
Try a tar instead of zip, that is a workaround for me for the weird zip behaviour in Galaxy.

**Comment by lecorguille** [comment]
Thanks for the reply but we address a community that is Windows-centric and that is easily stressed by a little shift in their practices :/

**Comment by lecorguille** [comment]
Gildas Le Corguillé @lecorguille 15:02
>My issue is that since this auto_decompression is set, my no_unzip_datatype fails.
So, now, the user have to select the type during the upload step. I still can’t see the point in this decompression :/

Marius van den Beek @mvdbeek 15:02
>The point used to be that at least you can upload a single compressed dataset and work with it
we do support more compressed datatypes now, but that is the reason

Gildas Le Corguillé @lecorguille 15:06
>Through the Galaxy UI, by default, data are decompress, until you specify zip. But it seems to me that this line is before the sniffing, isn’t it?

Marius van den Beek @mvdbeek 15:09
>Yes
if you want to check for multi-member archives and stop decompression at that point I think we can consider that

Gildas Le Corguillé @lecorguille 15:34
>Exactely, as explain in the issue, the dataset is complex :/
What are the risks to hard-disable this with a dirty patch?

Marius van den Beek @mvdbeek 15:36
>That you’ll need to figure this out again on the next release ;)
the handling isn’t great, if you have a suggestion we’ll (probably) take it

Gildas Le Corguillé @lecorguille 15:38
>I mean if I patch this line within usegalaxy.fr with auto_decompress = False in the meantime

Marius van den Beek @mvdbeek 15:40
>If someone uploads a dataset in a datatype that should be unpacked automatically it won’t be. Since that is different from other public Galaxy servers this is a bit of a problem
how much of a problem and if it is worth doing is difficult to say. It’s probably a better idea to open a pull-request with your intended changes and see what breaks
I’m sure there is a reasonable middleground for everyone, the sooner we find it the better

Gildas Le Corguillé @lecorguille 15:48
>My solution so far (but break this the data are decompress befor the sniffing): https://github.com/workflow4metabolomics/tools-metabolomics/blob/master/datatypes/datatype_no_unzip/no_unzip_datatypes.py#L16-L27

Marius van den Beek @mvdbeek 15:52
>To pedal back there a bit, does the zip datatype not work for you ?
if it does you may just want to add a sniffer with the logic you have in the NoUnzip class

Gildas Le Corguillé @lecorguille 15:58
>The sniff fails because the data are already uncompress.

Marius van den Beek @mvdbeek 15:59
>ok, that datatype doens’t have a sniffer
but if you just select that datatype it works fine for your purposes, right ?

Marius van den Beek @mvdbeek 16:13
>@lecorguille how about this:
```
diff --git a/lib/galaxy/datatypes/binary.py b/lib/galaxy/datatypes/binary.py
index 1f464bbdc9..f119b164d9 100644
--- a/lib/galaxy/datatypes/binary.py
+++ b/lib/galaxy/datatypes/binary.py
@@ -249,6 +249,11 @@ class CompressedZipArchive(CompressedArchive):
         except Exception:
             return "Compressed zip file (%s)" % (nice_size(dataset.get_size()))

+    def sniff(self, filename):
+        # If the zip file contains multiple files then return true
+        zf = zipfile.ZipFile(filename)
+        return len(zf.infolist()) > 1
+

 class GenericAsn1Binary(Binary):
     """Class for generic ASN.1 binary format"""
```
>that will result in zip archives in the history for multi-member zip files, with otherwise default Galaxy settings and no selection of the datatype.
We may need to tune the sniff order to make sure subclasses that implement their own sniffer sniff before this class, but I don’t think that’d be a major problem
(or it may just work ...)

Gildas Le Corguillé @lecorguille 16:19
>Great I will try that @mvdbeek
We can keep in touch in the issue

**Comment by lecorguille** [comment]
Hum, @mvdbeek, but it's still an issue if the data are uncompress before the sniffing.

It's what I get if I select no_unzip in the drop list :
```
Traceback (most recent call last):
  File "/shared/mfs/data/galaxy/server/tools/data_source/upload.py", line 326, in <module>
    __main__()
  File "/shared/mfs/data/galaxy/server/tools/data_source/upload.py", line 319, in __main__
    metadata.append(add_file(dataset, registry, output_path))
  File "/shared/mfs/data/galaxy/server/tools/data_source/upload.py", line 128, in add_file
    convert_spaces_to_tabs=dataset.space_to_tab,
  File "/shared/mfs/data/galaxy/server/lib/galaxy/datatypes/upload_util.py", line 70, in handle_upload
    if check_content and hasattr(datatype, 'sniff') and not datatype.sniff(path):
  File "/shared/mfs/data/galaxy/mutable-data/shed_tools/toolshed.g2.bx.psu.edu/repos/lecorguille/no_unzip_datatype/7800ba9a4c1e/no_unzip_datatype/no_unzip_datatypes.py", line 27, in sniff
    zf = zipfile.ZipFile(filename)
  File "/shared/mfs/data/software/miniconda/envs/_galaxy_/lib/python2.7/zipfile.py", line 770, in __init__
    self._RealGetContents()
  File "/shared/mfs/data/software/miniconda/envs/_galaxy_/lib/python2.7/zipfile.py", line 811, in _RealGetContents
    raise BadZipfile, "File is not a zip file"
zipfile.BadZipfile: File is not a zip file
```

**Comment by mvdbeek** [comment]
I think your misunderstanding, there is no need for `no_unzip ` with my patch.

**Comment by lecorguille** [comment]
Salut @mvdbeek 

Cool, it almost works. But I had to tweak it a little to deal with some hidden files: 

```
    def sniff(self, filename):
        # If the zip file contains multiple files then return true
        zf = zipfile.ZipFile(filename)
        zf_files = zf.infolist()
        # Filter the directories within the file list
        zf_files = [x.filename for x in zf_files if x.file_size > 0]
        # Filter hidden file within the file list
        hidden_file_patterns = ["__MACOSX",".DS_Store","^\."]
        regex = re.compile('|'.join(hidden_file_patterns))
        zf_files = [x for x in zf_files if x not in list(filter(regex.search, zf_files))]
        return len(zf_files) > 1
```

After a bunch of tests, we keep the original behaviour if there is a single file (even bury in a complex tree). 

What do you think about that? 
Do you think it worts a PR?

**Comment by mvdbeek** [comment]
Yes, please open a PR!

**Comment by nsoranzo** [comment]
Fixed in #9595.
